source/fab/steps/preprocess.py

```diff
@@ -0,0 +1,127 @@
+            - common_flags: Used to construct a :class:`~fab.config.FlagsConfig' object.
+            - path_flags: Used to construct a :class:`~fab.config.FlagsConfig' object.
+            - name: Used for logger output.
+
+        """
+        super().__init__(exe=preprocessor, common_flags=common_flags, path_flags=path_flags, name=name)
+
+        self.source_getter = source or self.DEFAULT_SOURCE
+        self.output_artefact = output_artefact or self.DEFAULT_OUTPUT_NAME
+        self.output_suffix = output_suffix or self.DEFAULT_OUTPUT_SUFFIX
+
+    def run(self, artefacts, config):
+        """
+        Preprocess all input files from `self.source_getter`, creating `self.output_artefact`.
+
+        This step uses multiprocessing, unless disabled in the :class:`~fab.steps.Step` class.
+
+        """
+        super().run(artefacts, config)
+
+        files = self.source_getter(artefacts)
+        results = self.run_mp(items=files, func=self.process_artefact)
+
+        # any errors?
+        exceptions = list(by_type(results, Exception))
+        if exceptions:
+            formatted_errors = "\n\n".join(map(str, exceptions))
+            raise Exception(
+                f"{formatted_errors}"
+                f"\n\n{len(exceptions)} "
+                f"Error(s) found during preprocessing: "
+            )
+
+        log_or_dot_finish(logger)
+        artefacts[self.output_artefact] = list(by_type(results, Path))
+
+    def process_artefact(self, fpath):
+        """
+        Expects an input file in the source folder.
+        Writes the output file to the output folder, with a lower case extension.
+
+        """
+        # output_fpath = self.output_path(fpath)
+        output_fpath = input_to_output_fpath(workspace=self._config.workspace, input_path=fpath).with_suffix(
+            self.output_suffix)
+
+        # for dev speed, but this could become a good time saver with, e.g, hashes or something
+        if self._config.debug_skip and output_fpath.exists():
+            log_or_dot(logger, f'Preprocessor skipping: {fpath}')
+            return output_fpath
+
+        if not output_fpath.parent.exists():
+            output_fpath.parent.mkdir(parents=True, exist_ok=True)
+
+        command = [self.exe]
+        command.extend(self.flags.flags_for_path(fpath, self._config.workspace))
+
+        # input and output files
+        command.append(str(fpath))
+        command.append(str(output_fpath))
+
+        log_or_dot(logger, 'PreProcessor running command: ' + ' '.join(command))
+        try:
+            run_command(command)
+        except Exception as err:
+            raise Exception(f"error preprocessing {fpath}: {err}")
+
+        return output_fpath
+
+
+class FortranPreProcessor(PreProcessor):
+    """
+    By default, preprocesses all .F90 and .f90 files in the *all_source* artefact,
+    creating the *preprocessed_fortran* artefact.
+
+    """
+    DEFAULT_SOURCE = FilterFpaths('all_source', ['.F90', '.f90'])
```

Comment:
Ah, interesting. I think we found some little Fs in one of the repos so added this here. We're currently just allowing everything to pass through the pre-processor, even if it's already pre-processed. We could instead identify pre-processed Fortran beforehand and not pass it through, with a small code change. That might require a little chat and possible code change, so are you happy for this to be spun off into a ticket?

Reply:
We don't expect to find little fs with pp directives. It's not clear if the feeling is to refactor this or just let them through, since it seems safe to do so, so we've [raised a ticket](https://github.com/metomi/fab/projects/8#card-79612822) for this.